### PR TITLE
NAS-104398 / 12.0 / Nas 104398

### PR DESCRIPTION
**Pull request by dkmullen**
Adds cores and threads to VM wizard, form and table; validates the three values against each other; changes vm form to ws; updates some values in vm wizard that have changed in mw for v. 12;

To save a new VM, I had to uncheck 'Enable VNC' on the first panel of the wizard; mw team is looking into this.

### CODE REVIEW

**Comment by erinix** [comment]
looks good to me, could you review the text @freenas/docs-team

File: src/app/helptext/vm/vm-wizard/vm-wizard.ts [hunk 1]
```diff
@@ -70,6 +69,18 @@ vcpus_tooltip: T('Number of virtual CPUs to allocate to the virtual\
  might also have operational or licensing\
  restrictions on the number of CPUs.'),
 
+cores: {
+  placeholder: T('Cores'),
+  tooltip: T('Specify the number of cores per virtual CPU socket. \
+ The product of vCPUs, cores and threads must not exceed 16.'),
```
Comment (a-stjohn) [inline src/app/helptext/vm/vm-wizard/vm-wizard.ts]:
```suggestion
 The product of vCPUs, cores, and threads must not exceed 16.'),
```

File: src/app/helptext/vm/vm-wizard/vm-wizard.ts [hunk 2]
```diff
@@ -70,6 +69,18 @@ vcpus_tooltip: T('Number of virtual CPUs to allocate to the virtual\
...
+}, 
+
+threads: {
+  placeholder: T('Threads'),
+  tooltip: T('Specify the number of threads per core. \
+ The product of vCPUs, cores and threads must not exceed 16.')
```
Comment (a-stjohn) [inline src/app/helptext/vm/vm-wizard/vm-wizard.ts]:
```suggestion
 The product of vCPUs, cores, and threads must not exceed 16.')
```

File: src/app/pages/vm/vm-form/vm-form.component.ts [hunk 3]
```diff
@@ -98,6 +167,27 @@ export class VmFormComponent {
...
+    return function validCPU(control: FormControl) {
+      const config = self.fieldConfig.find(c => c.name === name);
+        setTimeout(() => {
+          const errors = self.vcpus * self.cores * self.threads > 16
+          ? { validCPU : true }
+          : null;
+  
+          if (errors) {
+            config.hasErrors = true;
+            config.warnings = T(`The product of vCPUs, cores and threads must not exceed 16.`);
```
Comment (a-stjohn) [inline src/app/pages/vm/vm-form/vm-form.component.ts]:
```suggestion
            config.warnings = T(`The product of vCPUs, cores, and threads must not exceed 16.`);
```

**Pull request merged**